src/vm/eventpipe.cpp

```diff
@@ -367,10 +367,12 @@ void EventPipe::DisableInternal(EventPipeSessionID id, EventPipeProviderCallback
         else
         {
             _ASSERTE(!"Failed to get or create the EventPipeThread for rundown events.");
-            return;
```

Comment:
Why?

Reply:
Per Noah's comment above

Reply:
I'm guessing this should be fine since we no longer need pEventPipeThread to do rundown?

Reply:
If we hit that assert we have already failed to do rundown. The decision that remains at that point is whether or not we should do the rest of the cleanup below in the function (waiting on pending writes, flushing buffers to file, decreasing the count of active sessions, deleting the session). I think we should be doing those things.

Reply:
Maybe we should add a comment there. Usually, the assumption is that when you hit an assert, the rest will go south, which is probably not the case here.

Reply:
Can you change the comment a few lines above as follow?
`pSession->Disable(); // Suspend EventPipeBufferManager, and remove providers.`
to
`pSession->Disable(); // WriteAllBuffersToFile, and remove providers.`

Reply:
That begs the question, why write all buffers twice?

Reply:
The second `WriteAllBuffersToFile` happens to ensure that the rundown events get written before we go into the cleanup.  In theory that could happen inside the try/catch block.

Reply:
But why write it for the first time?

Reply:
To try and clear room in the buffers for rundown events.  Essentially giving rundown the best chance of getting through to the stream as we can without writing directly to the stream.

Reply:
I need to check if calling `WriteAllBuffersToFile` without calling `SuspendWriteEvents` first is safe. In particular, if writing is allowed, then all buffers is not a well-defined concept.

Reply:
I haven't picked apart this specific scenario, but I have a hunch that because the provider is already disabled, we will only see in-flight events causing new buffers to be allocated.  The old order was 
```
SuspendWriteEvent
WriteAllBuffersToFile
DisableAllProvidersForSession
```
and now it is
```
DisableAllProvidersForSesion
WriteAllBuffersToFile
EnableRundownProviders
ExecuteRundown
WriteAllBuffersToFile
```

You make a good point that we should check this ordering, so I'll take a look.

Reply:
Based on my reading, it looks like we should be okay to call `WriteAllBuffersToFile` since it marks the timestamp of the call and only writes events prior to that timestamp, see:
https://github.com/dotnet/coreclr/blob/234f1b6b238a9faa860a0bd1253d4baabdb35759/src/vm/eventpipebuffermanager.cpp#L643-L648

Regardless if there are new `WriteEvent` calls happening (there shouldn't be since we call `s_config.Disable(*pSession, pEventPipeProviderCallbackDataQueue);` first) we wouldn't be attempting to write any events written _after_ the call to `WriteAllBuffersToFile`.  We still do an actual `SuspendWriteEvent` before we go to clear the buffers out after rundown, so we shouldn't be running into a case where we have a writable buffer when we are doing cleanup.